deep_qa/training/models.py

```diff
@@ -146,6 +150,175 @@ def _make_predict_function(self):
+        else:
+            # May happen if we are running `fit` without Numpy input data,
+            # i.e. if all inputs to the models are data tensors
+            # instead of placeholders.
+            # In that case we will run `fit` over a single batch.
+            num_train_samples = batch_size
+            verbose = 2
+        index_array = numpy.arange(num_train_samples)
+
+        self.history = History()  # pylint: disable=attribute-defined-outside-init
+        callbacks = [BaseLogger()] + (callbacks or []) + [self.history]
+        if verbose:
+            callbacks += [ProgbarLogger()]
+        callbacks = CallbackList(callbacks)
+        out_labels = out_labels or []
+
+        # it's possible to callback a different model than self
+        # (used by Sequential models).
+        if hasattr(self, 'callback_model') and self.callback_model:
+            callback_model = self.callback_model
+        else:
+            callback_model = self  # pylint: disable=redefined-variable-type
+
+        callbacks.set_model(callback_model)
+        callbacks.set_params({
+                'batch_size': batch_size,
+                'epochs': epochs,
+                'samples': num_train_samples,
+                'verbose': verbose,
+                'do_validation': do_validation,
+                'metrics': callback_metrics or [],
+        })
+        callbacks.on_train_begin()
+        callback_model.stop_training = False
+        for cbk in callbacks:
+            cbk.validation_data = val_ins
+
+        for epoch in range(initial_epoch, epochs):
+            callbacks.on_epoch_begin(epoch)
+            if shuffle == 'batch':
+                index_array = _batch_shuffle(index_array, batch_size)
+            elif shuffle:
+                numpy.random.shuffle(index_array)
+
+            batches = _make_batches(num_train_samples, batch_size)
+            epoch_logs = {}
+            for batch_index, (batch_start, batch_end) in enumerate(batches):
+                batch_ids = index_array[batch_start:batch_end]
+                try:
+                    if isinstance(ins[-1], float):
+                        # Do not slice the training phase flag.
+                        ins_batch = _slice_arrays(ins[:-1], batch_ids) + [ins[-1]]
+                    else:
+                        ins_batch = _slice_arrays(ins, batch_ids)
+                except TypeError:
+                    raise TypeError('TypeError while preparing batch. '
+                                    'If using HDF5 input data, '
+                                    'pass shuffle="batch".')
+
+                # Here is the main difference between a single gpu model and one split
+                # across multiple gpus. In our multiple gpu model, all of the inputs
+                # are replicated num_gpus times, so we need to split our large batch
+                # into the corresponding sets of smaller batches for each model.
+                if hasattr(self, "num_gpus"):
```

Comment:
Checking this is a bit of a hack - do you think it would be better to add `self.num_gpus` as a variable which gets passed to compile?